akka-projection-slick/src/main/scala/akka/projection/slick/internal/SlickProjectionImpl.scala

```diff
@@ -0,0 +1,91 @@
+import akka.projection.slick.OffsetStore
+import akka.projection.{ Projection, ProjectionId }
+import akka.stream.scaladsl.{ Keep, Sink, Source }
+import akka.stream.{ KillSwitch, KillSwitches }
+import slick.basic.DatabaseConfig
+import slick.dbio.DBIO
+import slick.jdbc.JdbcProfile
+
+import scala.concurrent.{ ExecutionContext, Future, Promise }
+
+@InternalApi
+private[projection] class SlickProjectionImpl[Offset, StreamElement, P <: JdbcProfile](
+    val projectionId: ProjectionId,
+    sourceProvider: Option[Offset] => Source[StreamElement, _],
+    offsetExtractor: StreamElement => Offset,
+    databaseConfig: DatabaseConfig[P],
+    eventHandler: StreamElement => DBIO[Done])
+    extends Projection[StreamElement] {
+
+  private val offsetStore = new OffsetStore(databaseConfig.db, databaseConfig.profile)
+
+  private val promisedKillSwitch: Promise[KillSwitch] = Promise()
+  private val promiseToStop: Promise[Done] = Promise()
+
+  private val started = new AtomicBoolean(false)
+
+  override def start()(implicit systemProvider: ClassicActorSystemProvider): Unit = {
+
+    if (started.compareAndSet(false, true)) {
+      implicit val dispatcher: ExecutionContext = systemProvider.classicSystem.dispatcher
+
+      // TODO: add a LogSource for projection when we have a name and key
+      val akkaLogger = Logging(systemProvider.classicSystem, this.getClass)
+
+      val lastKnownOffset: Future[Option[Offset]] = offsetStore.readOffset(projectionId)
+
+      val futSource = lastKnownOffset.map { offsetOpt =>
+        akkaLogger.debug("Starting projection [{}] from offset [{}]", projectionId, offsetOpt)
+        sourceProvider(offsetOpt).viaMat(KillSwitches.single)(Keep.right)
+      }
+
+      val (killSwitch, stopped) =
+        Source.futureSource(futSource).mapAsync(1) { elt => processElement(elt) }.toMat(Sink.ignore)(Keep.both).run()
+
+      promisedKillSwitch.completeWith(killSwitch)
+      promiseToStop.completeWith(stopped)
+    }
+
+  }
+
+  override def stop()(implicit ec: ExecutionContext): Future[Done] = {
+    if (started.get()) {
+      promisedKillSwitch.future.flatMap { kill =>
+        kill.shutdown()
+        promiseToStop.future
+      }
+    } else {
+      Future.failed(new IllegalStateException(s"Projection [$projectionId] not started yet!"))
+    }
+  }
+
+  /**
+   * The method wrapping the user EventHandler function.
+   *
+   * @return A [[scala.concurrent.Future]] that represents the asynchronous completion of the user EventHandler
+   *         function.
+   */
+  override def processElement(elt: StreamElement)(implicit ec: ExecutionContext): Future[StreamElement] = {
+    import databaseConfig.profile.api._
+
+    // run user function and offset storage on the same transaction
+    // any side-effect in user function is at-least-once
+    val txDBIO =
+      offsetStore.saveOffset(projectionId, offsetExtractor(elt)).flatMap(_ => eventHandler(elt))
+
+    databaseConfig.db.run(txDBIO.transactionally).map(_ => elt)
```

Comment:
I think this will translate a `DBIOAction.failed` into a `Future` `Failure`, but I'm not sure. I don't think the existing test cases exercise this.

See https://github.com/akka/akka-projection/pull/35#discussion_r411661993